File: apps/scala/src/test/scala/machinelearning/chapter09/Samples.scala
package machinelearning.chapter09

import machinelearning.chapter02.{Row, Table}
import machinelearning.chapter02.Features

/** テストで使う架空の特徴量と表。 */
object Samples:

  /** RM の 1 列だけの特徴量。 */
  def rm(value: Double): Features = Features(Vector("RM"), Vector(value))

  /** 列名と値の組から 1 行を作る。 */
  def row(cells: (String, String)*): Row = Row(cells.toMap)

  /** 列名の並びと行から表を作る。 */
  def table(columns: Vector[String], rows: Row*): Table = Table(columns, rows.toVector)

File: apps/csharp/tests/MachineLearning.Tests/Chapter09/Samples.cs
namespace MachineLearning.Tests.Chapter09;

using MachineLearning.Chapter02;

/// <summary>第 9 章のテストで使う、架空の値の表と特徴量。</summary>
public static class Samples
{
    /// <summary>列名と、セルの並びから表を作る。</summary>
    public static Table Table(IReadOnlyList<string> columns, params string[][] rows)
    {
        ArgumentNullException.ThrowIfNull(columns);
        ArgumentNullException.ThrowIfNull(rows);
        return new Table(columns, [.. rows.Select(cells => Row(columns, cells))]);
    }

    /// <summary>列名と値の並びから 1 行を作る。</summary>
    public static Row Row(IReadOnlyList<string> columns, IReadOnlyList<string> cells)
    {
        ArgumentNullException.ThrowIfNull(columns);
        ArgumentNullException.ThrowIfNull(cells);
        return new Row(columns.Select((column, i) => (Column: column, Cell: cells[i]))
            .ToDictionary(pair => pair.Column, pair => pair.Cell, StringComparer.Ordinal));
    }

    /// <summary>1 列の特徴量を値の数だけ作る。</summary>
    public static IReadOnlyList<Features> Column(string name, params double[] values)
    {
        ArgumentNullException.ThrowIfNull(values);
        return [.. values.Select(value => new Features([name], [value]))];
    }
}
